Core/include/Acts/Utilities/Any.hpp

```diff
@@ -0,0 +1,427 @@
+        "Type needs to be move assignable and move constructible");
+    static_assert(
+        std::is_copy_assignable_v<U> && std::is_copy_constructible_v<U>,
+        "Type needs to be copy assignable and copy constructible");
+
+    cache.m_handler = makeHandler<U>();
+    if constexpr (sizeof(U) <= SIZE) {
+      // construct into local buffer
+      /*U* ptr =*/new (cache.m_data.data()) U(std::forward<Args>(args)...);
+    } else {
+      // too large, heap allocate
+      U* heap = new U(std::forward<Args>(args)...);
+      *reinterpret_cast<U**>(cache.m_data.data()) = heap;
+    }
+
+    return cache;
+  }
+
+  template <typename T, typename... Args>
+  explicit AnyBase(std::in_place_type_t<T> /*unused*/, Args&&... args) {
+    using U = std::decay_t<T>;
+    static_assert(std::is_same_v<T, std::decay_t<T>>,
+                  "Please pass the raw type, no const or ref");
+    static_assert(
+        std::is_move_assignable_v<U> && std::is_move_constructible_v<U>,
+        "Type needs to be move assignable and move constructible");
+    static_assert(
+        std::is_copy_assignable_v<U> && std::is_copy_constructible_v<U>,
+        "Type needs to be copy assignable and copy constructible");
+
+    m_handler = makeHandler<U>();
+    if constexpr (sizeof(U) <= SIZE) {
+      // construct into local buffer
+      /*U* ptr =*/new (m_data.data()) U(std::forward<Args>(args)...);
+    } else {
+      // too large, heap allocate
+      U* heap = new U(std::forward<Args>(args)...);
+      *reinterpret_cast<U**>(m_data.data()) = heap;
+    }
+  }
+
+#if defined(_ENABLE_ACTS_ANY_DEBUG)
+  AnyBase() { _ACTS_ANY_DEBUG("Default construct this=" << this); };
+#else
+  AnyBase() = default;
+#endif
+
+  template <typename T>
+  explicit AnyBase(T&& value) {
+    using U = std::decay_t<T>;
+    static_assert(
+        std::is_move_assignable_v<U> && std::is_move_constructible_v<U>,
+        "Type needs to be move assignable and move constructible");
+    static_assert(
+        std::is_copy_assignable_v<U> && std::is_copy_constructible_v<U>,
+        "Type needs to be copy assignable and copy constructible");
+
+    m_handler = makeHandler<U>();
+
+    if constexpr (sizeof(U) <= SIZE) {
+      // construct into local buffer
+      /*U* ptr =*/new (m_data.data()) U(std::move(value));
+      _ACTS_ANY_DEBUG(
+          "Construct local (this=" << this << ") at: " << (void*)m_data.data());
+    } else {
+      // too large, heap allocate
+      U* heap = new U(std::move(value));
+      _ACTS_ANY_DEBUG("Construct heap (this=" << this << ") at: " << heap);
+      _ACTS_ANY_DEBUG_BUFFER("-> buffer before", m_data);
+      *reinterpret_cast<U**>(m_data.data()) = heap;
+      _ACTS_ANY_DEBUG_BUFFER("-> buffer after", m_data);
+    }
+  }
+
+  template <typename T>
+  T& as() {
+    static_assert(std::is_same_v<T, std::decay_t<T>>,
+                  "Please pass the raw type, no const or ref");
+    assert(makeHandler<T>() == m_handler && "Bad type access");
```

Comment:
Are we sure we want the check only in an assert, and not e.g. with an exception? Seems quite dangerous to me...
Additionally that would also allow this behaviour to be unit-testable...

Reply:
Can switch to an exception